db/compaction/compaction_picker.cc

```diff
@@ -39,32 +39,46 @@ bool FindIntraL0Compaction(const std::vector<FileMetaData*>& level_files,
                            size_t min_files_to_compact,
                            uint64_t max_compact_bytes_per_del_file,
                            uint64_t max_compaction_bytes,
-                           CompactionInputFiles* comp_inputs) {
-  size_t compact_bytes = static_cast<size_t>(level_files[0]->fd.file_size);
-  uint64_t compensated_compact_bytes = level_files[0]->compensated_file_size;
+                           CompactionInputFiles* comp_inputs,
+                           SequenceNumber earliest_mem_seqno) {
+
+  // Do not pick ingested file when there is at least one memtable not flushed which of seqno is overlap with the sst.
+  size_t start = 0;
+  while (start < level_files.size() && level_files[start]->being_compacted &&
+      level_files[start]->fd.smallest_seqno == level_files[start]->fd.largest_seqno &&
+      level_files[start]->fd.largest_seqno >= earliest_mem_seqno) {
+    start ++;
+  }
+  if (start >= level_files.size()) {
+    return false;
+  }
+  size_t compact_bytes = static_cast<size_t>(level_files[start]->fd.file_size);
+  uint64_t compensated_compact_bytes = level_files[start]->compensated_file_size;
   size_t compact_bytes_per_del_file = port::kMaxSizet;
   // Compaction range will be [0, span_len).
   size_t span_len;
   // Pull in files until the amount of compaction work per deleted file begins
   // increasing or maximum total compaction size is reached.
   size_t new_compact_bytes_per_del_file = 0;
-  for (span_len = 1; span_len < level_files.size(); ++span_len) {
+  for (span_len = start + 1; span_len < level_files.size(); ++span_len) {
```

Comment:
I think the statement below should be
```
new_compact_bytes_per_del_file = compact_bytes / (span_len - start);
```

Reply:
Good catch. `span_len` isn't really the span length anymore, it's the exclusive end index of the span. Maybe we should call it `limit` or something.